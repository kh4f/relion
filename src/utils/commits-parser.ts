import type { CompleteCommitsParser, ParsedCommit, RawCommit, RawReference, RefLabel, Reference, Contributor, CommitMessage, CommitRange } from '@/types'
import { GpgSigLabel } from '@/enums'
import { getRawCommits } from '@/utils'
import { createHash } from 'node:crypto'

const parsedCommitsCache: Record<string, ParsedCommit> = {}
let recentReleaseTag: ParsedCommit['releaseTag']

export const parseCommits = (arg1: CommitRange | RawCommit[], commitsParser: CompleteCommitsParser, prevReleaseTagPattern: RegExp): ParsedCommit[] => {
	const rawCommits = Array.isArray(arg1) ? arg1 : getRawCommits(arg1, prevReleaseTagPattern)
	const parser = commitsParser

	const parsedCommits = rawCommits.map(commit => parseCommit(commit, parser, prevReleaseTagPattern))
		.filter(commit => commit !== null)

	recentReleaseTag = undefined

	return parsedCommits
}

export const parseCommit = (commit: RawCommit, parser: CompleteCommitsParser, prevReleaseTagPattern: RegExp): ParsedCommit | null => {
	if (typeof commit === 'string') commit = { message: commit }

	const { tagRefs, hash = getFakeCommitHash(commit.message) } = commit

	if (hash in parsedCommitsCache) return parsedCommitsCache[hash]

	const message = commit.message.trim()
	if (!message) throw new Error(`Message is missing for commit: ${JSON.stringify(commit)}`)

	let parsedMessage
	try {
		parsedMessage = parseCommitMessage(message, parser)
	} catch (error) {
		console.warn(`Error parsing commit '${hash}':`, (error as Error).message)
		return null
	}
	const { type, scope, subject, body, breakingChanges, footer } = parsedMessage
	const tags = tagRefs ? [...tagRefs.matchAll(parser.tagPattern)].map(m => m.groups?.tag ?? '') : []

	const signers = footer
		? [...footer.matchAll(parser.signerPattern)].map(m => m.groups as unknown as Contributor)
		: []

	const authors: Contributor[] = []
	const addAuthor = (contributor: Contributor): void => {
		if (!authors.some(a => a.email === contributor.email)) authors.push(contributor)
	}

	const author = commit.authorName && commit.authorEmail
		? getContributorDetails({ name: commit.authorName, email: commit.authorEmail }, signers)
		: undefined
	if (author) addAuthor(author)

	const committer = commit.committerName && commit.committerEmail
		? getContributorDetails({ name: commit.committerName, email: commit.committerEmail }, signers)
		: undefined
	if (committer) addAuthor(committer)

	const coAuthors = footer
		? [...footer.matchAll(parser.coAuthorPattern)].map(m => m.groups as unknown as Contributor)
			.map(coAuthor => getContributorDetails(coAuthor, signers))
		: []
	coAuthors.forEach(coAuthor => addAuthor(coAuthor))

	const refs = parseRefs((footer ?? ''), parser)

	const gpgSig = commit.gpgSigCode
		? {
			code: commit.gpgSigCode,
			label: GpgSigLabel[commit.gpgSigCode],
			keyId: commit.gpgSigKeyId,
		}
		: undefined

	let date = commit[parser.dateSource === 'committerDate' ? 'committerTs' : 'authorTs']
	if (typeof date === 'string') date = formatDate(new Date(+date * 1000), parser.dateFormat)

	const releaseTag = tags.find(tag => prevReleaseTagPattern.exec(tag))

	const associatedReleaseTag = releaseTag ?? recentReleaseTag
	if (associatedReleaseTag) recentReleaseTag = associatedReleaseTag

	const parsedCommit = { hash, type, scope, subject, body, breakingChanges, footer, committer, gpgSig, date, releaseTag, associatedReleaseTag,
		tags: tags.length ? tags : undefined,
		authors: authors.length ? authors : undefined,
		refs: refs.length ? refs : undefined,
	}

	if (hash && !(hash in parsedCommitsCache)) parsedCommitsCache[hash] = parsedCommit

	return parsedCommit
}

const getFakeCommitHash = (message: string): string =>
	'fake_' + createHash('sha256').update(message, 'utf8').digest('hex').slice(0, 7)

const parseCommitMessage = (message: string, parser: CompleteCommitsParser): CommitMessage => {
	const [header, ...details] = message.split('\n\n')

	const headerMatch = parser.headerPattern.exec(header)
	if (!headerMatch?.groups) throw new Error(`Commit header '${header}' doesn't match expected format`)
	const { type, scope, bang, subject } = headerMatch.groups

	let breakingChanges
	const breakingChangesPart = details.find(detail => parser.breakingChangesPattern.test(detail))
	if (breakingChangesPart) {
		breakingChanges = parseBreakingChanges(breakingChangesPart, parser)
		details.splice(details.indexOf(breakingChangesPart), 1)
	} else if (bang) {
		breakingChanges = subject
	}

	const footerStart = details.findIndex(detail =>
		detail.match(parser.refActionPattern)
		?? detail.match(parser.coAuthorPattern)
		?? detail.match(parser.signerPattern))
	const [body, footer] = footerStart === -1
		? [details.join('\n\n'), '']
		: [details.slice(0, footerStart).join('\n\n'), details.slice(footerStart).join('\n\n')]

	return {
		type,
		scope: scope || undefined,
		subject,
		body: body || undefined,
		breakingChanges,
		footer: footer || undefined,
	}
}

const parseBreakingChanges = (value: string, parser: CompleteCommitsParser): ParsedCommit['breakingChanges'] => {
	const breakingChanges = parser.breakingChangesPattern.exec(value)?.groups?.content
	if (!breakingChanges) throw new Error(`Failed to extract breaking changes content from '${value}' using pattern "${parser.breakingChangesPattern}"`)

	const breakingChangeList = [...breakingChanges.matchAll(parser.breakingChangeListPattern)]

	return breakingChangeList.length
		? breakingChangeList.map(m => m[1])
		: breakingChanges
}

const getContributorDetails = (contributor: Contributor, signers: Contributor[]): Contributor => {
	const hasSignedOff = signers.some(signer => signer.email === contributor.email && signer.name === contributor.name)
	return {
		...contributor,
		hasSignedOff,
		ghLogin: contributor.name,
		ghUrl: `https://github.com/${contributor.name}`,
	}
}

const parseRefs = (value: string, parser: CompleteCommitsParser): Reference[] =>
	[...value.matchAll(parser.refPattern)].map(m => m.groups as unknown as RawReference)
		.filter(rawRef => parser.refActionPattern.test(rawRef.action))
		.flatMap(rawRef =>
			[...rawRef.labels.matchAll(parser.refLabelPattern)].map(m => m.groups as unknown as RefLabel)
				.filter(label => !!label.number)
				.map(label => ({
					action: rawRef.action,
					owner: label.owner,
					repo: label.repo,
					number: label.number,
				})),
		)

const formatDate = (date: Date, format: string): string => {
	const pad = (num: number) => num.toString().padStart(2, '0')

	if (format === 'US') {
		return date.toLocaleString('en-US', { month: 'short', day: 'numeric', year: 'numeric' })
	} else if (format === 'ISO') {
		return date.toISOString().split('T')[0]
	}

	const dateParts: Record<string, string> = {
		YYYY: date.getUTCFullYear().toString(),
		MM: pad(date.getUTCMonth() + 1),
		DD: pad(date.getUTCDate()),
		HH: pad(date.getUTCHours()),
		mm: pad(date.getUTCMinutes()),
		ss: pad(date.getUTCSeconds()),
	}

	return format.replace(/YYYY|MM|DD|HH|mm|ss/g, k => dateParts[k])
}