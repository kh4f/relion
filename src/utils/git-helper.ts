import { execSync } from 'node:child_process'
import type { CommitRange, RawCommit, RepoInfo } from '@/types'
import { defaultConfig } from '@/defaults'

const commitLogFormat = `##COMMIT##%n#HASH# %h%n#MSG# %B%n#REFS# %d%n#AUTHOR-NAME# %an%n#AUTHOR-EMAIL# %ae%n#AUTHOR-DATE# %at%n#COMMITTER-NAME# %cn%n#COMMITTER-EMAIL# %ce%n#COMMITTER-DATE# %ct%n#GPGSIG-CODE# %G?%n#GPGSIG-KEYID# %GK%n`
const rawCommitPattern = /##COMMIT##\n#HASH# (?<hash>.+)?\n#MSG# (?<message>[\s\S]*?)\n#REFS#\s+(?<tagRefs>.+)?\n#AUTHOR-NAME# (?<authorName>.+)?\n#AUTHOR-EMAIL# (?<authorEmail>.+)?\n#AUTHOR-DATE# (?<authorTs>.+)?\n#COMMITTER-NAME# (?<committerName>.+)?\n#COMMITTER-EMAIL# (?<committerEmail>.+)?\n#COMMITTER-DATE# (?<committerTs>.+)?\n#GPGSIG-CODE# (?<gpgSigCode>.+)?\n#GPGSIG-KEYID# (?<gpgSigKeyId>.+)?/g

const releaseTagsCache = new Map<string, string[]>()

export const getRepoInfo = (remoteUrlPattern: RegExp): RepoInfo => {
	const remoteUrl = execSync('git remote get-url origin', { encoding: 'utf8' }).trim()
	const remoteUrlMatch = remoteUrlPattern.exec(remoteUrl)
	if (!remoteUrlMatch?.groups) throw new Error(`Couldn't parse remote URL: ` + remoteUrl)

	const { host, owner, name } = remoteUrlMatch.groups
	const homepage = `https://${host}/${owner}/${name}`

	return { host, owner, name, homepage }
}

export const getRawCommits = (commitRange: CommitRange, prevReleaseTagPattern?: RegExp): RawCommit[] => {
	const firstCommitHash = getFirstCommitHash()
	const releaseTags = getReleaseTags(prevReleaseTagPattern)

	let from = '', to = '', range = ''
	if (commitRange === 'all') {
		from = '{{firstCommit}}'
		to = 'HEAD'
	} else if (commitRange === 'unreleased') {
		from = releaseTags[0] ?? '{{firstCommit}}'
		to = 'HEAD'
	} else if (commitRange === 'latest-release') {
		from = releaseTags[1] ?? '{{firstCommit}}'
		to = releaseTags[0] ?? 'HEAD'
	} else if (typeof commitRange === 'object') {
		const targetTagIndex = releaseTags.indexOf(commitRange.releaseTag)
		if (targetTagIndex === -1) throw new Error(`Release tag '${commitRange.releaseTag}' not found`)
		from = releaseTags[targetTagIndex + 1] ?? '{{firstCommit}}'
		to = releaseTags[targetTagIndex]
	} else {
		range = commitRange
	}

	if (from && to) range = from === '{{firstCommit}}' ? `"${from}^!" ${to}` : `"${from}..${to}"`

	range = range.replace('{{firstCommit}}', firstCommitHash)

	const gitLogCommits = execSync(`git log ${range} --pretty="${commitLogFormat}"`, { encoding: 'utf8' })
	return [...gitLogCommits.matchAll(rawCommitPattern)].map(m => m.groups as unknown as RawCommit)
}

const getFirstCommitHash = (): string => execSync('git rev-list --max-parents=0 HEAD', { encoding: 'utf8' }).trim()

export const getReleaseTags = (tagPattern?: RegExp): string[] => {
	tagPattern = tagPattern ?? defaultConfig.prevReleaseTagPattern
	const cacheKey = tagPattern.toString()
	if (releaseTagsCache.has(cacheKey)) return releaseTagsCache.get(cacheKey) ?? []

	const rawTags = execSync('git tag --sort=-creatordate', { encoding: 'utf8' })
	const tags = rawTags.split('\n').filter(tag => tagPattern.test(tag))
	releaseTagsCache.set(cacheKey, tags)
	return tags
}