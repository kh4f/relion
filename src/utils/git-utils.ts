import { execSync } from 'node:child_process'
import type { CommitRange, RawCommit, RepoInfo } from '@/types'

const commitLogFormat = `##COMMIT##%n#HASH# %h%n#MSG# %B%n#REFS# %d%n#AUTHOR-NAME# %an%n#AUTHOR-EMAIL# %ae%n#AUTHOR-DATE# %at%n#COMMITTER-NAME# %cn%n#COMMITTER-EMAIL# %ce%n#COMMITTER-DATE# %ct%n#GPGSIG-CODE# %G?%n#GPGSIG-KEYID# %GK%n`
const rawCommitPattern = /##COMMIT##\n#HASH# (?<hash>.+)?\n#MSG# (?<message>[\s\S]*?)\n#REFS#\s+(?<tagRefs>.+)?\n#AUTHOR-NAME# (?<authorName>.+)?\n#AUTHOR-EMAIL# (?<authorEmail>.+)?\n#AUTHOR-DATE# (?<authorTs>.+)?\n#COMMITTER-NAME# (?<committerName>.+)?\n#COMMITTER-EMAIL# (?<committerEmail>.+)?\n#COMMITTER-DATE# (?<committerTs>.+)?\n#GPGSIG-CODE# (?<gpgSigCode>.+)?\n#GPGSIG-KEYID# (?<gpgSigKeyId>.+)?/g

export const getRepoInfo = (remoteUrlPattern: RegExp): RepoInfo => {
	const remoteUrl = execSync('git remote get-url origin', { encoding: 'utf8' }).trim()
	const remoteUrlMatch = remoteUrlPattern.exec(remoteUrl)
	if (!remoteUrlMatch?.groups) throw new Error(`Couldn't parse remote URL: ` + remoteUrl)

	const { host, owner, name } = remoteUrlMatch.groups
	const homepage = `https://${host}/${owner}/${name}`

	return { host, owner, name, homepage }
}

export const getRawCommits = (commitRange: CommitRange, prevReleaseTagPattern: RegExp): RawCommit[] => {
	const firstCommitHash = getFirstCommitHash()
	const versionTags = getVersionTags(prevReleaseTagPattern)

	let from: string, to: string
	if (commitRange === 'all') {
		from = firstCommitHash
		to = 'HEAD'
	} else if (commitRange === 'unreleased') {
		from = versionTags[0] ?? firstCommitHash
		to = 'HEAD'
	} else if (commitRange === 'latest-release') {
		from = versionTags[1] ?? firstCommitHash
		to = versionTags[0] ? versionTags[0] + '^' : 'HEAD'
	} else if ('from' in commitRange || 'to' in commitRange) {
		const fromValue = 'from' in commitRange ? commitRange.from : 'firstCommit'
		from = fromValue === 'firstCommit' ? firstCommitHash : fromValue
		to = 'to' in commitRange ? commitRange.to : 'HEAD'
	} else if ('versionTag' in commitRange) {
		const targetTagIndex = versionTags.indexOf(commitRange.versionTag)
		if (targetTagIndex === -1) throw new Error(`Version tag '${commitRange.versionTag}' not found`)
		from = versionTags[targetTagIndex]
		to = versionTags[targetTagIndex + 1] ?? 'HEAD'
	} else {
		throw new Error(`Invalid commit range provided`)
	}

	const gitLogCommits = execSync(`git log ${from}..${to} --pretty="${commitLogFormat}"`, { encoding: 'utf8' })
	return [...gitLogCommits.matchAll(rawCommitPattern)].map(m => m.groups as RawCommit)
}

const getFirstCommitHash = (): string => execSync('git rev-list --max-parents=0 HEAD', { encoding: 'utf8' }).trim()

export const getVersionTags = (tagPattern: RegExp): string[] => {
	const rawTags = execSync('git tag --sort=-creatordate', { encoding: 'utf8' })
	return rawTags.split('\n').filter(tag => tagPattern.test(tag))
}