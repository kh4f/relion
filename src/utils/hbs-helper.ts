import Handlebars from 'handlebars'
import releaseTemplate from '@/templates/release.hbs'

export const compilePartials = (partials: Record<string, string | ((fallback: string) => string)>): Record<string, HandlebarsTemplateDelegate> => {
	const resolvedPartials: Record<string, string> = Object.fromEntries(Object.entries(partials).map(([key, templateOrFunction]) => [key, typeof templateOrFunction === 'function' ? transformFallback(key, templateOrFunction) : templateOrFunction]))
	return Object.fromEntries(Object.entries(resolvedPartials).map(([key, template]) => [key, Handlebars.compile(template)]))
}

const transformFallback = (partialName: string, transform: (fallback: string) => string): string => {
	const fallback = new RegExp(`${partialName}.*?}}\\s*(.*?){{~?/${partialName}`, 's').exec(releaseTemplate)?.[1]
	if (!fallback) throw new Error(`Partial "${partialName}" not found in the release template.`)
	return transform(fallback)
}