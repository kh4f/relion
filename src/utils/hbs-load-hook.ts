import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import type { LoadHook } from 'node:module'

export const load: LoadHook = (url, context, nextLoad) => url.endsWith('.hbs')
	? {
		format: 'module',
		shortCircuit: true,
		source: `export default ${JSON.stringify(readFileSync(fileURLToPath(url), 'utf8'))}`,
	}
	: nextLoad(url, context)