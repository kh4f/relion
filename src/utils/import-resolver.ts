import { dirname, resolve as pathResolve } from 'node:path'
import { fileURLToPath, pathToFileURL } from 'node:url'
import type { ResolveHook } from 'node:module'

export const resolve: ResolveHook = (specifier, context, defaultResolve) => {
	if (!/^\.{1,2}\//.test(specifier) || !context.parentURL) return defaultResolve(specifier, context)
	const parentDir = dirname(fileURLToPath(context.parentURL))
	const absPath = pathResolve(parentDir, specifier + '.ts')
	const fileUrl = pathToFileURL(absPath).href
	return defaultResolve(fileUrl, context)
}