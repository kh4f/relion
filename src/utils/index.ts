export * from './config-resolver'
export * from './version-manager'
export * from './commits-parser'
export * from './git-utils'
export * from './logger'

import { inspect } from 'node:util'

export const deepLog = (value: unknown, depth?: number): void => {
	console.log(inspect(value, { depth: depth ?? null, colors: true }))
}