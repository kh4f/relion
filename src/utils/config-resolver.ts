import { parseVersion, determineNextVersion, getReleaseTags, getRepoInfo, parseCommits, parseCommit, renderTemplate, extractVersionFromTag, compilePartials, log } from '@/utils'
import type { UserConfig, ResolvedConfig, TransformedConfig, VersionedFile, MergedConfig, FalseOrComplete, ParsedCommit, ReleaseWithFlatCommits, ReleaseWithTypeGroups, TypeGroupsMap, ResolvedCommit, FilledTypeGroupMap, ScopeGroup } from '@/types'
import { defaultConfig, defaultVersionedFiles, defaultChangelogOptions, defaultCommitOptions, defaultTagOptions } from '@/defaults'

export const resolveConfig = (userConfig: UserConfig): ResolvedConfig => {
	const profileMergedConfig = mergeProfileConfig(userConfig)
	const mergedConfig = mergeWithDefaults(profileMergedConfig)
	const transformedConfig = transformVersionedFiles(mergedConfig)
	const ResolvedConfig = resolveContext(transformedConfig)
	const finalConfig = resolveTemplates(ResolvedConfig)
	return finalConfig
}

const mergeProfileConfig = (baseConfig: UserConfig): UserConfig => {
	const profileName = baseConfig.profile
	if (!profileName) return baseConfig

	const profileConfig = baseConfig[`_${profileName}`]
	if (!profileConfig) throw new Error(`Profile "${profileName}" not found in configuration.`)

	const mergeOption = <T extends keyof UserConfig>(propKey: T, ...nestedPropKeys: string[]): UserConfig[T] => {
		type PlainObject = Record<string, unknown>

		const isPlainObject = (value: unknown): value is PlainObject =>
			Object.prototype.toString.call(value) === '[object Object]'

		const mergeObjects = (baseObj: unknown, overrideObject: unknown): unknown => {
			const isBasePlainObject = isPlainObject(baseObj)
			const isOverridePlainObject = isPlainObject(overrideObject)
			if (isBasePlainObject && isOverridePlainObject) return { ...baseObj, ...overrideObject }
			return overrideObject ?? baseObj
		}

		const baseConfigProp = baseConfig[propKey] as PlainObject | undefined
		const profileConfigProp = profileConfig[propKey] as PlainObject | undefined
		const result = mergeObjects(baseConfigProp, profileConfigProp) as PlainObject | undefined
		if (result === undefined) return undefined

		nestedPropKeys.forEach((key) => {
			result[key] = mergeObjects(baseConfigProp?.[key], profileConfigProp?.[key])
		})

		return result as UserConfig[T]
	}

	return {
		...baseConfig, ...profileConfig,
		commitsParser: mergeOption('commitsParser'),
		changelog: mergeOption('changelog', 'partials', 'helpers'),
		commit: mergeOption('commit'),
		tag: mergeOption('tag'),
		context: mergeOption('context'),
	}
}

const mergeWithDefaults = (userConfig: UserConfig): MergedConfig => {
	const resolveOptions = <T>(optionsName: string, options: boolean | T | undefined, defaults: Required<T>, ...subObjects: (keyof T)[]): FalseOrComplete<T> => {
		if (options == undefined || options === false) return false
		if (options === true) return defaults
		if (typeof options !== 'object') throw new Error(`Invalid value for ${optionsName}. It should be a boolean or an object.`)
		const result: Required<T> = { ...defaults, ...options }
		subObjects.forEach(subObjectKey => result[subObjectKey] = { ...defaults[subObjectKey], ...options[subObjectKey] })
		return result
	}

	return {
		...defaultConfig, ...userConfig,
		commitsParser: { ...defaultConfig.commitsParser, ...userConfig.commitsParser },
		changelog: resolveOptions('changelog', userConfig.changelog, defaultChangelogOptions, 'partials', 'helpers'),
		commit: resolveOptions('commit', userConfig.commit, defaultCommitOptions),
		tag: resolveOptions('tag', userConfig.tag, defaultTagOptions),
	}
}

const transformVersionedFiles = (config: MergedConfig): TransformedConfig => {
	const resolveVersionedFile = (filePath: string): VersionedFile => {
		const matchingDefaultVersionFile = defaultVersionedFiles.find(defaultFile =>
			defaultFile.filePathRegex.test(filePath))
		if (matchingDefaultVersionFile) {
			return { filePath, versionPattern: matchingDefaultVersionFile.versionPattern }
		} else {
			throw new Error(
				`File ${filePath} doesn't match any default versioned files. `
				+ 'Please provide a custom version pattern for this file.',
			)
		}
	}

	const resolveBump = (bump: MergedConfig['bump']): false | VersionedFile[] => {
		if (bump === false) return false
		if (bump === true) return [versionSourceFile]
		if (Array.isArray(bump)) return [
			versionSourceFile,
			...(bump.map(bumpFile =>
				typeof bumpFile === 'string' ? resolveVersionedFile(bumpFile) : bumpFile,
			)),
		]
		throw new Error('Invalid value for bump. It should be a boolean or an array.')
	}

	const versionSourceFile = typeof config.versionSourceFile === 'string'
		? resolveVersionedFile(config.versionSourceFile)
		: config.versionSourceFile

	return {
		...config,
		versionSourceFile,
		bump: resolveBump(config.bump),
		changelog: config.changelog === false
			? false
			: { ...config.changelog, compiledPartials: compilePartials(config.changelog.partials) },
	}
}

const resolveContext = (config: TransformedConfig): ResolvedConfig => {
	const oldContext = config.context ?? {}

	const repoInfo = getRepoInfo(config.commitsParser.remoteUrlPattern)
	const currentVersion = oldContext.currentVersion ?? (
		config.versionSource === 'latest-release-tag'
			? extractVersionFromTag(getReleaseTags(config.prevReleaseTagPattern)[0], config.prevReleaseTagPattern) ?? '0.0.0'
			: parseVersion(config.versionSourceFile)
	)
	log(`Current version (from ${config.versionSource === 'latest-release-tag' ? 'latest release tag' : config.versionSourceFile.filePath}): '${currentVersion}'`)
	const currentTag = oldContext.currentTag ?? getReleaseTags(config.prevReleaseTagPattern)[0]
	const newVersion = oldContext.newVersion ?? determineNextVersion(config, currentVersion)
	const newTag = oldContext.newTag ?? (config.newTagPrefix
		? config.newTagPrefix + newVersion
		: config.newTagFormat.replace('{{version}}', newVersion))

	const commitRange = config.changelog ? config.changelog.commitRange : 'unreleased'
	const parsedCommits = oldContext.commits
		? oldContext.commits.map((commit) => {
			return ((typeof commit === 'object' && 'message' in commit) || typeof commit === 'string')
				? parseCommit(commit, config.commitsParser, config.prevReleaseTagPattern)
				: commit
		}).filter(c => c != null)
		: parseCommits(commitRange, config.commitsParser, config.prevReleaseTagPattern)

	const resolvedCommits = resolveCommits(parsedCommits, newTag, config.commitsParser.revertCommitBodyPattern)
	const releases = config.changelog ? groupCommitsByReleases(resolvedCommits, config.changelog.sections, config.prevReleaseTagPattern, config.changelog.groupCommitsByScope) : null
	const { commits: _, ...noCommitsOldContext } = oldContext

	return {
		...config,
		context: {
			currentVersion,
			currentTag,
			newVersion,
			newTag,
			commits: resolvedCommits,
			releases,
			...noCommitsOldContext,
			repo: {
				...repoInfo,
				...config.context?.repo,
			},
		},
	}
}

const resolveCommits = (commits: ParsedCommit[], newTag: string, revertCommitBodyPattern: RegExp): ResolvedCommit[] => {
	let breakingChangesIndex = 0
	return commits.map((commit) => {
		let isRevertedStatus: ResolvedCommit['isReverted'] = null
		const revertCommit = commits.find(c => c.type === 'revert' && revertCommitBodyPattern.exec(c.body ?? '')?.groups?.hash === commit.hash)
		if (revertCommit) isRevertedStatus = revertCommit.associatedReleaseTag === commit.associatedReleaseTag ? 'inTheSameRelease' : 'inOtherRelease'
		return {
			...commit,
			associatedReleaseTag: commit.associatedReleaseTag ?? newTag,
			isReverted: commit.isReverted ?? isRevertedStatus,
			breakingChangeIndex: commit.breakingChanges ? ++breakingChangesIndex : undefined,
		}
	})
}

const groupCommitsByReleases = (commits: ResolvedCommit[], sections: TypeGroupsMap, prevReleaseTagPattern: RegExp, withScopeGroups?: boolean): ReleaseWithTypeGroups[] => {
	const releases: Record<string, ReleaseWithFlatCommits> = {}

	commits.forEach((commit) => {
		const releaseTag = commit.associatedReleaseTag
		if (releaseTag in releases) {
			releases[releaseTag].commits.push(commit)
		} else {
			releases[releaseTag] = {
				tag: releaseTag,
				version: extractVersionFromTag(releaseTag, prevReleaseTagPattern),
				date: commit.date,
				commits: [commit],
			}
		}
	})

	return Object.values(releases)
		.map(({ commits, ...rest }) => ({ ...rest, commitTypeGroups: groupCommitsByType(commits, sections, withScopeGroups) }))
		.filter(release => Object.keys(release.commitTypeGroups).length)
}

const groupCommitsByType = (commits: ResolvedCommit[], sections: TypeGroupsMap, withScopeGroups?: boolean): FilledTypeGroupMap => {
	const filledTypeGroupsMap: FilledTypeGroupMap = Object.fromEntries(
		Object.entries(sections).map(([id, { filter: _, ...defWithoutFilter }]) => [id, { ...defWithoutFilter, commits: [] }]),
	)

	commits.forEach((commit) => {
		const isBreaking = !!commit.breakingChanges
		let isGrouped = false
		let isBreakingGrouped = false

		for (const sectionId in sections) {
			if (sections[sectionId].filter && !sections[sectionId].filter(commit)) continue

			const sectionTypes = [sections[sectionId].commitType].flat()

			if (isBreaking && !isBreakingGrouped && sectionTypes.includes('breaking')) {
				filledTypeGroupsMap[sectionId].commits.push(commit)
				isBreakingGrouped = true
				continue
			}

			if (!isGrouped && (sectionTypes.includes(commit.type) || sectionTypes.includes('*'))) {
				filledTypeGroupsMap[sectionId].commits.push(commit)
				isGrouped = true
			}

			if (isGrouped && (!isBreaking || isBreakingGrouped)) return
		}
	})

	const typeGroupsEntries = Object.entries(filledTypeGroupsMap).filter(([_, group]) => group.commits.length)
	return Object.fromEntries(
		withScopeGroups
			? typeGroupsEntries.map(([type, group]) => [type, { ...group, scopeGroups: groupCommitsByScope(group.commits) }])
			: typeGroupsEntries,
	)
}

const groupCommitsByScope = (commits: ResolvedCommit[]): ScopeGroup[] => {
	const groups: Record<string, ResolvedCommit[]> = {}
	commits.forEach((commit) => {
		const scope = commit.scope ?? ''
		groups[scope] ??= []
		groups[scope].push(commit)
	})
	return Object.entries(groups).map(([scope, commits]) => ({ scope, commits }))
}

const resolveTemplates = (config: ResolvedConfig): ResolvedConfig => {
	return {
		...config,
		commit: config.commit
			? { ...config.commit, message: renderTemplate(config.commit.message, config.context) }
			: config.commit,
		tag: config.tag
			? {
				...config.tag,
				name: renderTemplate(config.tag.name, config.context),
				message: renderTemplate(config.tag.message, config.context),
			}
			: config.tag,
	}
}