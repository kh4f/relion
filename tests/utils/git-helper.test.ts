import { describe, expect, it } from 'vitest'
import { getRawCommits } from '@/utils/git-helper'

describe('getRawCommits with commitsScope', () => {
	it('should use cwd as default scope when not provided', () => {
		const hashes = getRawCommits('v0.18.0..v0.19.0').map(c => c.hash)
		expect(hashes).toMatchSnapshot()
	})

	it('should filter commits by specific path', () => {
		const hashes = getRawCommits('v0.18.0..v0.19.0', undefined, 'src/utils').map(c => c.hash)
		expect(hashes).toMatchSnapshot()
	})

	it('should filter commits by multiple paths', () => {
		const hashes = getRawCommits('v0.18.0..v0.19.0', undefined, 'src tests').map(c => c.hash)
		expect(hashes).toMatchSnapshot()
	})
})