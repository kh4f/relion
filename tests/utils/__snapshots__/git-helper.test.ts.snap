// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getRawCommits with commitsScope > should filter commits by multiple paths 1`] = `
[
  "693492a",
  "0cf7ae2",
  "10bfbf9",
  "8b2b221",
  "68ef725",
  "32f7221",
  "c46e531",
  "bf8371f",
  "4001441",
  "037ebd8",
  "d2e5287",
  "d569b3d",
  "a59963f",
  "0c6cf75",
  "2f17c70",
  "4e4f71f",
  "79274ca",
  "be9204c",
  "aa08d6a",
  "c46e41b",
  "3132b13",
  "866f177",
  "747f953",
  "9054b8a",
  "c1f942a",
  "80eeecf",
  "2b8c32c",
  "011abc0",
  "8bbb9be",
  "8ad16fd",
  "a934ed4",
]
`;

exports[`getRawCommits with commitsScope > should filter commits by specific path 1`] = `
[
  "0cf7ae2",
  "10bfbf9",
  "8b2b221",
  "c46e531",
  "037ebd8",
  "0c6cf75",
  "2f17c70",
  "79274ca",
  "c46e41b",
  "c1f942a",
  "8ad16fd",
  "a934ed4",
]
`;

exports[`getRawCommits with commitsScope > should use cwd as default scope when not provided 1`] = `
[
  "6d4412d",
  "cb1e944",
  "2aa65d9",
  "693492a",
  "0cf7ae2",
  "10bfbf9",
  "8b2b221",
  "68ef725",
  "32f7221",
  "c46e531",
  "bf8371f",
  "4001441",
  "037ebd8",
  "d2e5287",
  "d569b3d",
  "a59963f",
  "0c6cf75",
  "2f17c70",
  "4e4f71f",
  "79274ca",
  "be9204c",
  "aa08d6a",
  "c46e41b",
  "3132b13",
  "866f177",
  "747f953",
  "9054b8a",
  "c1f942a",
  "80eeecf",
  "2b8c32c",
  "011abc0",
  "5bd64fe",
  "8bbb9be",
  "750c55d",
  "d399c4f",
  "079b693",
  "8ad16fd",
  "51a8366",
  "0515831",
  "003e291",
  "a934ed4",
  "1d98fa2",
]
`;
